fastscanner/services/indicators/tests/test_subscribe_realtime.py

```diff
@@ -354,12 +389,128 @@ async def test_channel_handler_multiple_indicators(setup):
     assert "prev_day_close" in row
     assert "daily_gap" in row
     assert "cumulative_daily_volume" in row
+
+
+@pytest.mark.asyncio
+async def test_interval_calculation():
+    handler = TestHandler()
+
+    channel_handler_1min = CandleChannelHandler(
+        "AAPL", [], handler, "1min", candle_timeout=0.1
+    )
+    assert channel_handler_1min._expected_count == 1
+
+    channel_handler_3min = CandleChannelHandler(
+        "AAPL", [], handler, "3min", candle_timeout=0.1
+    )
+    assert channel_handler_3min._expected_count == 3
+
+    channel_handler_1h = CandleChannelHandler(
+        "AAPL", [], handler, "1h", candle_timeout=0.1
+    )
+    assert channel_handler_1h._expected_count == 60
+
+    channel_handler_2h = CandleChannelHandler(
+        "AAPL", [], handler, "2h", candle_timeout=0.1
+    )
+    assert channel_handler_2h._expected_count == 120
+
+
+@pytest.mark.asyncio
+async def test_candle_aggregation(setup):
+    symbol, ts, handler = setup
+
+    channel_handler = CandleChannelHandler(
+        symbol, [], handler, "3min", candle_timeout=0.2
+    )
+
+    ts_start = pd.Timestamp(ts).floor("3min")
+    ts1 = ts_start
+    ts2 = ts_start + timedelta(minutes=1)
+    ts3 = ts_start + timedelta(minutes=2)
+
+    msg1 = create_stream_message(ts1, open=100, high=105, low=98, close=102, volume=500)
+    msg2 = create_stream_message(
+        ts2, open=102, high=108, low=101, close=107, volume=700
+    )
+    msg3 = create_stream_message(
+        ts3, open=107, high=110, low=105, close=108, volume=300
+    )
+
+    mock_clock = MagicMock()
+    ClockRegistry.set(mock_clock)
+    mock_clock.now.return_value = ts_start
+
+    try:
+        await channel_handler.handle(f"candles_min_{symbol}", msg1)
+        await channel_handler.handle(f"candles_min_{symbol}", msg2)
+        await channel_handler.handle(f"candles_min_{symbol}", msg3)
+
+        await asyncio.sleep(0.05)
+
+        assert len(handler.received) == 1
+        _, row = handler.received[0]
+
+        assert row[CandleCol.OPEN] == 100
+
+        assert row[CandleCol.HIGH] == 110
+
+        assert row[CandleCol.LOW] == 98
+
+        assert row[CandleCol.CLOSE] == 108
+
+        assert row[CandleCol.VOLUME] == 1500
+
+        assert row.name == ts_start
+    finally:
+        ClockRegistry.unset()
+
+
+@pytest.mark.asyncio
+async def test_flush_on_timeout_with_partial_buffer(setup):
+    symbol, ts, handler = setup
+    channel_handler = CandleChannelHandler(
+        symbol, [], handler, "3min", candle_timeout=0.2
+    )
+    ts_start = pd.Timestamp(ts).floor("3min")
+
+    msg1 = create_stream_message(
+        ts_start, open=100, high=101, low=99, close=100, volume=100
+    )
+    msg2 = create_stream_message(
+        ts_start + timedelta(minutes=1),
+        open=100,
+        high=102,
+        low=98,
+        close=101,
+        volume=200,
+    )
+
+    mock_clock = MagicMock()
+    ClockRegistry.set(mock_clock)
+    mock_clock.now.return_value = ts_start
```

Comment:
Here's where you have to set the clock to `ts_start + timedelta(minutes=2, seconds=59, microseconds=990000)`. So that the timeout is properly set.